### PR TITLE
Guarantee viewable on ready

### PR DESCRIPTION
**Pull request by robwalch**
Cleans up view and setup order and uses brute force check for visibility if IntersectionObserver does not callback before ready.

JW7-4064
JW7-2009 (retest autostart="viewable")

### CODE REVIEW

**Comment by robwalch** [comment]
@jw-kaurand 
```
IE 11.0.0 (Windows 7 0.0.0)
Unspecified error.
  at /var/lib/jenkins-slave/workspace/jw7-opensource-unit-tests/node_modules/intersection-observer/intersection-observer.js:622
```

File: src/js/controller/controller.js [hunk 1]
```diff
@@ -280,33 +274,119 @@ define([
...
                 }
             }
 
-            function _onActiveTabChange() {
-                _checkVisibility();
-            }
+            function _getVisibility() {
+                // Set visibility to 1 if we're in fullscreen
+                var fullscreen = _model.get('fullscreen');
+                if (fullscreen) {
```
Comment (johnBartos) [inline src/js/controller/controller.js]: `if (_model.get('fullscreen'))`

File: src/js/controller/controller.js [hunk 2]
```diff
@@ -280,33 +274,119 @@ define([
...
+                    return 1;
+                }
 
-            function _checkVisibility() {
                 // Set visibility to 0 if we're not in the active tab
+                var activeTab = _model.get('activeTab');
```
Comment (johnBartos) [inline src/js/controller/controller.js]: `if (!_model.get('activeTab'))`

File: src/js/controller/controller.js [hunk 3]
```diff
@@ -280,33 +274,119 @@ define([
...
                 // Otherwise, set it to the intersection ratio reported from the intersection observer
                 var intersectionRatio = _model.get('intersectionRatio');
-                var activeTab = _model.get('activeTab');
 
-                // We don't yet know if the player is viewable
-                if (activeTab && _.isUndefined(intersectionRatio)) {
-                    return;
+                if (_.isUndefined(intersectionRatio)) {
+                    // Set visibility to 1 if we're in an iFrame and intersection is unknown
+                    var iFrame = _model.get('iFrame');
```
Comment (johnBartos) [inline src/js/controller/controller.js]: `if (_model.get('iFrame'))`

File: src/js/controller/controller.js [hunk 4]
```diff
@@ -280,33 +274,119 @@ define([
...
+                    // Get intersectionRatio through brute force
+                    intersectionRatio = _computeVisibility(_view.element());
                 }
 
-                var visibility = activeTab ? intersectionRatio : 0;
-                _model.set('visibility', visibility);
+                return intersectionRatio;
+            }
+
+            function _computeVisibility(target) {
```
Comment (johnBartos) [inline src/js/controller/controller.js]: How hard would it be to put this function in a util so we avoid +90 LOC in the controller?

File: src/js/controller/controller.js [hunk 5]
```diff
@@ -280,33 +274,119 @@ define([
...
+                    top: top,
+                    bottom: bottom,
+                    left: left,
+                    right: right,
+                    width: width,
+                    height: height
+                };
+            }
+
+            function _setViewable(visibility) {
```
Comment (johnBartos) [inline src/js/controller/controller.js]: why does a set also function as a get?

Comment (robwalch) [inline src/js/controller/controller.js]: It simplifies `_onVisibilityChange` so we don't have to read off the model after setting it.

Comment (robwalch) [inline src/js/controller/controller.js]: @johnBartos I'm changing it to an update functions which simplifies everything.

**Pull request merged**